phase transition: average phase value

lhy11009 opened the pull request:
Pull Request Checklist. Please read and check each box with an X. Delete any part not applicable. Ask on the [forum](https://community.geodynamics.org/c/aspect) if you need help with any step.

This is the second step of implementing phase transitions to the material model. I added a phase_average_value function to utilities.cc calculate parameters of compositions from the parameters of phases using the gamma function. On top of that, I changed the implementation of function phase_average_equation_of_state_outputs in interfaces.cc.
As @gassmoeller has addressed before, we should not call this phase_average_value function if there is no phase on a composition. That's what I have taken care of in this PR. Although I haven't included the viscosity part of it, which I intend to deal with in the third part. 
Thanks, @gassmoeller for review the last one, this one and another one in the near future -v-

### Before your first pull request:

* [x] I have read the guidelines in our [CONTRIBUTING.md](../blob/master/CONTRIBUTING.md) document.

### For all pull requests:

* [x] I have followed the [instructions for indenting my code](../blob/master/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/master/tests/) directory.
* [x] I have added a changelog entry in the [doc/modules/changes](../blob/master/doc/modules/changes) directory that will inform other users of my change.

gassmoeller commented on include/aspect/material_model/equation_of_state/interface.h:
```diff
@@ -113,8 +113,7 @@ namespace aspect
     phase_average_equation_of_state_outputs(const EquationOfStateOutputs<dim> &eos_outputs_all_phases,
-                                            const MaterialUtilities::PhaseFunction<dim> &phase_function,
-                                            MaterialUtilities::PhaseFunctionInputs<dim> &phase_in,
+                                            const std::pair<std::vector<double>, const std::vector<unsigned int>> &gamma_inputs,
```
You are right that this interface is not ideal, because it would compute the phase function values several times (once for the equation of state, once for the rheology, and so on). But your alternative is somewhat confusing, because `gamma_inputs` does not explain what it is. What about the following interface:

```
    phase_average_equation_of_state_outputs(const EquationOfStateOutputs<dim> &eos_outputs_all_phases,
                                            const std::vector<double> &phase_function_values, 
                                            const std::vector<unsigned int>> &n_phases_per_composition,
                                            EquationOfStateOutputs<dim> &eos_outputs);
```

Please also update the documentation test above this function.

gassmoeller commented on include/aspect/material_model/utilities.h:
```diff
@@ -268,6 +268,29 @@ namespace aspect
+      * to the other. And this is achieved by the definition of gamma function, which varies
+      * between 0 and 1.
+      */
+      double phase_average_value (const std::pair<std::vector<double>, const std::vector<unsigned int>> &gamma_inputs,
+                                  const std::vector<double> &parameter_values,
+                                  const unsigned int composition,
+                                  bool by_log=false);
```
Same comment as above, I would suggest a different interface to make things easier to understand. What about the following:

```suggestion
      double phase_average_value (const std::vector<double> &parameter_values,
                                  const std::vector<double> &phase_function_values, 
                                  const std::vector<unsigned int> &n_phases_per_composition,
                                  const unsigned int composition_index,
                                  const MaterialAveraging::AveragingOperation operation = MaterialAveraging::arithmetic_average);
```

See the comment inside the function about how to use the `MaterialAveraging::AveragingOperation`.

gassmoeller commented on source/material_model/equation_of_state/interface.cc:
```diff
@@ -43,47 +43,23 @@ namespace aspect
-              eos_outputs.entropy_derivative_pressure[c] = gamma * (eos_outputs_all_phases.entropy_derivative_pressure[j]-eos_outputs_all_phases.entropy_derivative_pressure[j-1]);
-              eos_outputs.entropy_derivative_temperature[c] = gamma * (eos_outputs_all_phases.entropy_derivative_temperature[j]-eos_outputs_all_phases.entropy_derivative_temperature[j-1]);
-            }
+          eos_outputs.densities[c] = \
```
Yes I like this. Much simpler than before. Please remove the `\` though, C++ only ends statements when it encounters an `;` anyway.

gassmoeller commented on source/material_model/utilities.cc:
```diff
@@ -699,6 +699,43 @@ namespace aspect
 
 
+      // Calculate a parameter of single composition at single point by averaging phases.
+      // Either linear(by_log is 0) or using log form(by_log is 1), default is linear.
```
Remove the comment above. Documentation about the general purpose of this function belongs in the header file.

gassmoeller commented on source/material_model/utilities.cc:
```diff
@@ -699,6 +699,43 @@ namespace aspect
+                                  const unsigned int composition,
+                                  bool by_log)
+      {
+        const std::vector<double> &gamma_values = gamma_inputs.first;
+        const std::vector<unsigned int> &n_phases_per_composition = gamma_inputs.second;
```
Since you take this parameter apart immediately when you enter the function you can as well hand it over as two separate parameters, which is easier to read.

gassmoeller commented on source/material_model/utilities.cc:
```diff
@@ -699,6 +699,43 @@ namespace aspect
+        if (n_phases_per_composition[composition] > 0)
+          {
+            // Do averaging when there are multiple phases
+            if (by_log)
```
```suggestion
            if (operation == MaterialAveraging::log_average)
```

gassmoeller commented on source/material_model/utilities.cc:
```diff
@@ -699,6 +699,43 @@ namespace aspect
+            // Do averaging when there are multiple phases
+            if (by_log)
+              averaged_parameter = log(averaged_parameter);
+            for (unsigned int i=0; i<n_phases_per_composition[composition]; ++i)
```
add a newline before the `for` to make clear that line 720 is the only line inside the `if (log)`.

gassmoeller commented on source/material_model/utilities.cc:
```diff
@@ -699,6 +699,43 @@ namespace aspect
+            for (unsigned int i=0; i<n_phases_per_composition[composition]; ++i)
+              {
+                Assert(base+i+1<parameter_values.size(), ExcInternalError());
+                if (by_log)
```
same as above

gassmoeller commented on source/material_model/utilities.cc:
```diff
@@ -699,6 +699,43 @@ namespace aspect
+                    // This is used for averaging prefactors in flow laws.
+                    averaged_parameter += gamma_values[base-composition+i] * log(parameter_values[base+i+1] / parameter_values[base+i]);
+                  }
+                else
```
```suggestion
                else if (operation == MaterialAveraging::arithmetic_average)
                   averaged_parameter += gamma_values[base-composition+i] * (parameter_values[base+i+1] - parameter_values[base+i]);
                else
                  AssertThrow(false, ExcInternalError());
```

lhy11009 replied:
@gassmoeller Hi, Rene, I think I have addressed all of your comments, with expect that I have to name the variable ar phase_transition_arithmetic in PhaseAveragingOperation in order to not conflict with the one defined in CompositionalAveragingOperation, is there a better solution to that?

lhy11009 replied:
And right now, I see the first checking failed with the cookbook and some issues with make which I don't quite understand.

gassmoeller commented on source/material_model/visco_plastic.cc:
```diff
@@ -464,6 +464,13 @@ namespace aspect
 
       std::vector<double> average_elastic_shear_moduli (in.n_evaluation_points());
 
+      // Store value of gamma functions, and number of gamma function for each compostion  # todo
+      // Construct a gamma_inputs structure for passing that information into functions
+      // While number of gamma function is set from the start, value of gamma functions keep updating on every point
+      std::vector<double> gamma_values(phase_function.n_phase_transitions(), 0.0);
+      std::pair<std::vector<double>, const std::vector<unsigned int>> gamma_inputs =
+                                                                     std::make_pair(gamma_values, phase_function.n_phase_transitions_for_each_composition());
```
```suggestion
      // Store value of phase function for each phase and composition
      // While the number of phases is fixed, the value of the phase function is updated for every point
      std::vector<double> phase_function_values(phase_function.n_phase_transitions(), 0.0);
```

gassmoeller commented on source/material_model/visco_plastic.cc:
```diff
@@ -485,9 +492,17 @@ namespace aspect
           phase_average_equation_of_state_outputs(eos_outputs_all_phases,
-                                                  phase_function,
-                                                  phase_inputs,
+                                                  gamma_inputs,
```
```suggestion
                                                  phase_function_values,
                                                  phase_function.n_phase_transitions(),
```

gassmoeller commented:
/rebuild

gassmoeller commented:
Yes I see the problem. `MaterialAveraging` is defined inside `interface.h`, but that is not included in `utilities.h`, because `interface.h` already includes `utilities.h`. So we have a circular dependency here, mostly caused by the fact that we were not careful with the relation between interface and utilities. Solving this is outside the scope of this PR, so lets go a different route: We should probably create a different averaging operation for phases anyway, because it behaves differently than `MaterialAveraging`. Please add the following above `phase_average_value` inside `utilities.h` (this is a modified copy of the `CompositionalAveragingOperation` we already have):
```
      /**
       * For material models with multiple phases:
       * Enumeration for selecting which averaging scheme to use when
       * averaging the properties of different phases.
       * Select between arithmetic and logarithmic.
       */
      enum PhaseAveragingOperation
      {
        arithmetic,
        logarithmic
      };
```

Then you can use `PhaseAveragingOperation` as the type of your `operation` parameter, and `PhaseAveragingOperation::logarithmic`/`PhaseAveragingOperation::arithmetic` inside the function.

lhy11009 commented:
Hi, I commented a week ago. But maybe I should put it in a separate comment instead of putting it under one of the previous comments
@gassmoeller Hi, Rene, I think I have addressed all of your comments, except that I have to name the variable ar phase_transition_arithmetic in PhaseAveragingOperation in order to not conflict with the one defined in CompositionalAveragingOperation, is there a better solution to that?
@gassmoeller, And right now, I see the first checking failed with the cookbook and some issues with make which I don't quite understand.

gassmoeller commented on include/aspect/material_model/utilities.h:
```diff
@@ -268,6 +268,42 @@ namespace aspect
+      };
+
+      /**
+      * For multicomponent material models:
```
```suggestion
```

gassmoeller commented on include/aspect/material_model/utilities.h:
```diff
@@ -268,6 +268,42 @@ namespace aspect
+      * nonzero values at a given quadrature point. This means that properties
+      * for each composition have to be averaged based on the fractions of each
+      * phase field present. This function performs this type of averaging.
+      * The averaging is based on the choice in @p by_log. Averaging is conducted
```
```suggestion
      * The averaging is based on the choice in @p operation. Averaging is conducted
```

gassmoeller commented on include/aspect/material_model/utilities.h:
```diff
@@ -268,6 +268,42 @@ namespace aspect
+      * for each composition have to be averaged based on the fractions of each
+      * phase field present. This function performs this type of averaging.
+      * The averaging is based on the choice in @p by_log. Averaging is conducted
+      * over the gamma functions of phases given in @gamma_inputs.second, with
```
```suggestion
      * over the phase functions given in @phase_function_values, with
```

gassmoeller commented on include/aspect/material_model/utilities.h:
```diff
@@ -268,6 +268,42 @@ namespace aspect
+      * phase field present. This function performs this type of averaging.
+      * The averaging is based on the choice in @p by_log. Averaging is conducted
+      * over the gamma functions of phases given in @gamma_inputs.second, with
+      * parameter_values stand for values of all the phases. Unlike the average_value
```
```suggestion
      * @parameter_values containing values of all individual phases. Unlike the average_value
```

gassmoeller commented on include/aspect/material_model/utilities.h:
```diff
@@ -268,6 +268,42 @@ namespace aspect
+      * The averaging is based on the choice in @p by_log. Averaging is conducted
+      * over the gamma functions of phases given in @gamma_inputs.second, with
+      * parameter_values stand for values of all the phases. Unlike the average_value
+      * function defined for compositions, averaging in this function is calculated base
```
```suggestion
      * function defined for compositions, averaging in this function is calculated based
```

gassmoeller commented on include/aspect/material_model/utilities.h:
```diff
@@ -268,6 +268,42 @@ namespace aspect
+      * over the gamma functions of phases given in @gamma_inputs.second, with
+      * parameter_values stand for values of all the phases. Unlike the average_value
+      * function defined for compositions, averaging in this function is calculated base
+      * on gamma functions and the change of variables on the trajectory of phase boundaries.
```
```suggestion
      * on phase functions and the change of variables on the trajectory of phase boundaries.
```

gassmoeller commented on include/aspect/material_model/utilities.h:
```diff
@@ -268,6 +268,42 @@ namespace aspect
+      * parameter_values stand for values of all the phases. Unlike the average_value
+      * function defined for compositions, averaging in this function is calculated base
+      * on gamma functions and the change of variables on the trajectory of phase boundaries.
+      * Thus on a single phase boundary, values of variables change stepwise from one phase
```
```suggestion
      * Thus on a single phase boundary, values of variables change gradually from one phase
```

gassmoeller commented on include/aspect/material_model/utilities.h:
```diff
@@ -268,6 +268,42 @@ namespace aspect
+      * function defined for compositions, averaging in this function is calculated base
+      * on gamma functions and the change of variables on the trajectory of phase boundaries.
+      * Thus on a single phase boundary, values of variables change stepwise from one phase
+      * to the other. And this is achieved by the definition of gamma function, which varies
+      * between 0 and 1.
```
```suggestion
      * to the other. The values of the phase function used to average the properties varies
      * between 0 and 1.
```

lhy11009 commented:
That's okay. Happy movement! It's an event to be celebrated in my culture. I have addressed all your comments. The solution for the enum type does make sense, so there should be one namespace for a type. I have also created a change_log while I am not sure if it is needed.

gassmoeller commented on doc/modules/changes/20200701_haoyuanli:
```diff
@@ -0,0 +1,8 @@
+New: Added function phase_average_value, namespace PhaseUtilities to namespace MaterialModel::MaterialUtilities.
+This new function averages parameters(e.g density, viscosity) across all phases for a composition. It
+takes parameters of phases and values of gamma functions as inputs, applys an arithmetic or 
+logarithmic averaging and return the averaged value. A enum type PhaseUtilities::PhaseAveragingOperation 
+is defined to determine the type of operation(e.g arithmetic, logarithmic) to perform in the phase_average_value
+function.
```
Thanks for the changelog. We usually try to keep the description here as short as possible, the exact usage of the function can be seen in the documentation of the header file. This is more a note to users that this new functionality exists. Here is how I would change it:
```suggestion
New: Added a new function PhaseUtilities::phase_average_value to the namespace MaterialModel::MaterialUtilities.
This new function averages parameters(e.g density, viscosity) across all phases for a composition.
```

lhy11009 commented:
Yes, that makes sense. I have committed the change to the change_log
